### PR TITLE
Merge with_values when merging relations

### PR DESCRIPTION
**Pull request by joelvh**
CTE `WITH` statements get lost when merging relations. This ensures the ActiveRecord `Merger` keeps track of `with_values`.

### CODE REVIEW

File: lib/activerecord/cte/core_ext.rb
Comment (vlado) [inline lib/activerecord/cte/core_ext.rb:11]: Not sure why `normal_values` method is needed?

**Review by vlado — COMMENTED** [review]
Thanks @joelvh. Looks great! Left one suggestion and one question.

File: test/activerecord/cte_test.rb
Comment (joelvh) [reply test/activerecord/cte_test.rb]: @vlado when i was initially setting up the test, not selecting from the CTE's explicitly didn't trigger an error. However, it must have been something else that eventually solved itself. So, I've updated the test case to be much more intuitive and check size and the result in efc936e.

File: lib/activerecord/cte/core_ext.rb
Comment (joelvh) [reply lib/activerecord/cte/core_ext.rb:11]: @vlado my observation was that [`Merger` loops through a list of values](https://github.com/rails/rails/blob/v6.0.2.1/activerecord/lib/active_record/relation/merger.rb#L61) and this would cause it to implicitly follow the same process for `with`.

**Comment by joelvh** [comment]
@vlado i know you've been trying to get this merged into Rails. What do you think about making this compatible with Rails 6.1? The only test that fails for 6.1 is `test_with_when_string_is_passed_as_an_argument` -- I'm using this gem with Rails 6.1 because I'm not using strings, but thought it might be worth thinking about.

Thanks for getting CTE's working so succinctly with this gem!

**Comment by vlado** [comment]
> @vlado i know you've been trying to get this merged into Rails. What do you think about making this compatible with Rails 6.1? The only test that fails for 6.1 is `test_with_when_string_is_passed_as_an_argument` -- I'm using this gem with Rails 6.1 because I'm not using strings, but thought it might be worth thinking about.
> 
> Thanks for getting CTE's working so succinctly with this gem!

Haven't used gem with Rails 6.1 yet but it definitely makes sense to make it compatible. Looks like core team is interested to finally merge PR to the Rails actually https://github.com/rails/rails/pull/37944#issuecomment-850729200. I plan to update the PR and then port the changes to the gem also.

File: lib/activerecord/cte/core_ext.rb
Comment (vlado) [reply lib/activerecord/cte/core_ext.rb:11]: Got it now. I was looking at the main branch https://github.com/rails/rails/blob/main/activerecord/lib/active_record/relation/merger.rb that does not use `normal_values`. I think this is fine for this PR then and we can change / improve when adding 6.1 compatibility.

**Review by vlado — APPROVED** [review]

**Comment by joelvh** [comment]
Thenks @vlado! Looking forward to see where things go with adding this to Rails. Been keeping tabs for a while now. Happy to help where I can. Best of luck!

**Comment by dark-panda** [comment]
This change ended up breaking some queries for me. The basic gist of it is something like this:

```ruby
User.with(foo: 'select 1 = 1')
  .merge(
    User.with(foo: 'select 1 = 1')
  ).to_sql
```

In v0.1.1, this would produce the following:

```sql
WITH "foo" AS (select 1 = 1) SELECT "users".* FROM "users"
```

In v0.1.3, this would produce the following:

```sql
WITH "foo" AS (select 1 = 1), "foo" AS (select 1 = 1) SELECT "users".* FROM "users"
```

This errors out as you cannot have two `"foo"` `WITH` aliases in the same scope in the SQL query.

The question becomes: which is the desired behaviour? Should it error out when it sees about multiple definitions of `"foo"`? Should it overwrite the first `"foo"` with the second?

**Comment by joelvh** [comment]
@dark-panda i ran into this scenario which lead to the logic to make sure we have a uniq list of values. Maybe that can be improved more?

File: lib/activerecord/cte/core_ext.rb
Comment (joelvh) [inline lib/activerecord/cte/core_ext.rb:22]: @dark-panda this may need to be improved

**Comment by dark-panda** [comment]
@joelvh I think maybe this could be the issue:

```ruby
module WithMerger
    def normal_values
      super + %i[with]
    end
```

That is causing the base `#merge` method in `ActiveRecord::Relation::Merger#merge` to also merge `relation` with the values from `with_values`, and then after that we still merge them in `#merge_withs`, so it's like a double merge. If you comment out the `#normal_values` method, both of these are sort of correct:

```ruby
puts User.with(foo: 'select 1 = 1').merge(User.with(foo: 'select 1 = 1')).to_sql
WITH "foo" AS (select 1 = 1) SELECT "users".* FROM "users"

puts User.with(foo: 'select 1 = 1').merge(User.with(foo: 'select 2 = 2')).to_sql
WITH "foo" AS (select 1 = 1), "foo" AS (select 2 = 2) SELECT "users".* FROM "users"
```

I think this is the desired behaviour -- if you have two identical `WITH` statements it does not merge them, but if you have different ones, it does, but in that case I think you'd want an exception. This also fixes my application, which uses `WITH` pretty extensively and it doesn't break any of the unit tests.

**Comment by vlado** [comment]
@dark-panda Thanks for reporting this 👍 

Raising an error seems more logical to me. I see the use case for overwriting the first statement with the second one (based on some condition you want to change the original one) but I think it could lead to unexpected results more often than be useful.

**Comment by dark-panda** [comment]
@vlado yeah I think it should not raise an error if it’s the same statement but raise an error on repeats, or even just let it pass to the database as my suggested change above does. That would at the very least work with my use case, so I’d be happy at least. We use CTEs somewhat extensively in a few sections of our app.

**Comment by joelvh** [comment]
@dark-panda im also curious which version of Rails you're using. There are some slight differences between 6.0 and 6.1. I'm on 6.1 and don't recall if the change you're suggesting breaks 6.1.

Can you submit a PR?

**Comment by dark-panda** [comment]
@joelvh this was on 6.1.3.2.

**Comment by dark-panda** [comment]
@joelvh i’ll see if I can do a PR later on, shouldn’t be a problem.

**Comment by dark-panda** [comment]
@joelvh @vlado PR incoming, but one thing I noticed is that with Rails 6.1, the `test_with_when_string_is_passed_as_an_argument`. This happens regardless of my fix for this particular issue.

The failure is as follows:

```
Error:
Activerecord::CteTest#test_with_when_string_is_passed_as_an_argument:
Arel::Visitors::UnsupportedVisitError: Unsupported argument type: NilClass. Construct an Arel node instead.
    /Users/jay/Setup/ruby/rails-git/activerecord/lib/arel/visitors/to_sql.rb:712:in `unsupported'
    /Users/jay/Setup/ruby/rails-git/activerecord/lib/arel/visitors/visitor.rb:30:in `visit'
    ...
```

**Comment by dark-panda** [comment]
Pull request is here:

https://github.com/vlado/activerecord-cte/pull/3

Note that I left the failing test in there. This test succeeds in Rails 6.0 but fails in Rails 6.1.

**Comment by vlado** [comment]
> @joelvh @vlado PR incoming, but one thing I noticed is that with Rails 6.1, the `test_with_when_string_is_passed_as_an_argument`. This happens regardless of my fix for this particular issue.
> 
> The failure is as follows:
> 
> ```
> Error:
> Activerecord::CteTest#test_with_when_string_is_passed_as_an_argument:
> Arel::Visitors::UnsupportedVisitError: Unsupported argument type: NilClass. Construct an Arel node instead.
>     /Users/jay/Setup/ruby/rails-git/activerecord/lib/arel/visitors/to_sql.rb:712:in `unsupported'
>     /Users/jay/Setup/ruby/rails-git/activerecord/lib/arel/visitors/visitor.rb:30:in `visit'
>     ...
> ```

Yes, I'm aware that passing string as a CTE query is broken with Rails 6.1. Will need to add ActiveRecord 6.1 to the [test matrix](https://github.com/vlado/activerecord-cte/blob/master/bin/test#L9-L12) and submit the fix :)

**Pull request merged**
